draft-irtf-cfrg-hybrid-kems.md

```diff
@@ -303,236 +425,231 @@ MUST provide the bit-security required to source input randomness for PQ/T
-specified in {{QSF}}.  The KitchenSink construction is maximally conservative
-in design, opting for the least assumptions about the component KEMs. The QSF
-construction is tailored to specific component KEMs and is not generally
-reusable; specific requirements for component KEMs to be usable in the QSF
-combiner are detailed in {{QSF}}.
-
-Both make use of the following requirements:
-
-1. Both component KEMs have IND-CCA security.
-2. KDF as a secure PRF. A key derivation function (KDF) that is modeled as a
-secure pseudorandom function (PRF) in the standard model {{GHP2018}} and
-independent random oracle in the random oracle model (ROM).
-3. Fixed-length values. Every instantiation in concrete parameters of the
-generic constructions is for fixed parameter sizes, KDF choice, and label,
-allowing the lengths to not also be encoded into the generic
-construction. The label/KDF/component algorithm parameter sets MUST be
-disjoint and non-colliding. Moreover, the length of each each public
-encapsulation key, ciphertext, and shared secret is fixed once the algorithm
-is assumed to be fixed.
-
-## Key generation and derivation {#generic-keygen}
-
-For both constructions in this document we provide a common key generation
-and derivation design. It relies on the following parameters that are
-populated by concrete instantiations:
-
-* `XOF`: the eXtended Output Function
-* `PQKEM`: the PQ KEM component scheme
-* `G`: the nomimal group used to construct the traditional KEM component
-  scheme as described in {{group}}
-* Nseed: length in bytes of the seed randomness sourced from the RNG
-* Npqseed: length in bytes of the input to PQ.DeriveKeyPair()
-* Ntradseed: length in bytes of the input to NominalGroup.ScalarFromBytes()
+## PreHashedKeys {#prehash}
 
-~~~
-def expandDecapsulationKey(dk):
-  expanded = XOF(dk, Npqseed + Ntradseed)
-  (pq_EK, pq_DK) = PQKEM.DeriveKeyPair(expanded[..Npqseed])
-  trad_DK = G.ScalarFromBytes(expanded[Npqseed..])
-  trad_EK = G.SerializeElement(NominalGroup.ScalarMultBase(trad_DK))
-  return (pq_DK, trad_DK, pq_EK, trad_EK)
+The PreHashedKeys hybrid KEM is a performance optimization of the HashEverything
+KEM, optimized for the case where encapsulation keys are large and frequently
+reused. In such cases, hashing the entire encapsulation key is expensive, and
+the same value is hashed repeatedly.  The PreHashedKeys KEM thus computes an
+intermediate hash of the hybrid encapsulation key, so that the hash value can be
+computed once and used across many encapsulation or decapsulation operations.
 
-def GenerateKeyPair():
-  dk = random(Nseed)
-  (pq_DK, trad_DK, pq_EK, trad_EK) = expandDecapsulationKey(dk)
-  return dk, concat(pq_EK, trad_EK)
-~~~
+The PreHashedKeys KEM is identical to the HashEverything KEM except for the
+shared secret computation.  One additional KDF is required:
 
-Similarly, `DeriveKeyPair` works as follows:
+* `KeyHash`: A KDF producing byte strings of length `KEM_H.Nss` (`KeyHash.Nout
```

Comment:
`KeyHash` does not have to have the same requirements of a KDF, it probably only needs collision resistance (but I'm not sure because we don't have a proof! @pag-crypto )